daemon/container.go

```diff
@@ -155,11 +156,22 @@ func (daemon *Daemon) newContainer(name string, operatingSystem string, config *
 	base.Args = args // FIXME: de-duplicate from config
 	base.Config = config
 	base.HostConfig = &containertypes.HostConfig{}
-	base.ImageID = imgID
 	base.NetworkSettings = &network.Settings{IsAnonymousEndpoint: noExplicitName}
 	base.Name = name
 	base.Driver = daemon.imageService.GraphDriverName()
 	base.OS = operatingSystem
+	if img != nil {
+		base.ImageID = img.ID()
+		// TODO do we need a "fallback" here for images that have empty OS/Arch/etc (in which case, use runtime.GOARCH?)
+		//      would be nice if img had methods for this with that implemented (img.OS(), img.Architecture(), img.Platform())
+		base.Platform = ocispec.Platform{
```

Comment:
Do we want to "normalize" this information here, or keep it "as-is" and handle it on the client side? https://github.com/containerd/containerd/blob/v1.5.2/platforms/platforms.go#L265-L278

Reply:
I added containerd's `Normalize` code here, but perhaps this is something to be discussed.

Reply:
IMO doing the `normalize` client-side makes sense (or doing the `normalize` before we ever store/use the data in the first place :innocent:)

Reply:
I originally started with that (do the normalising on the client side), then noticed that "normalising" is doing more than just changing casing and (e.g.) `x86 -> amd64`. It's also filling in missing information based on the current platform it's running on (which obviously wouldn't work if the client is running on a different platform as the daemon.

From that, I was also wondering if the client should just _present_ the information the daemon provides (in most situations i think that's the ideal), hence I moved it here.

> or doing the normalize before we ever store/use the data in the first place

You mean "when we store the image" ? That felt a bit tricky; should we modify image data?

Reply:
Arg, sorry, didn't realize this was straight from the image.  In that case I'd be even more in favor of the API passing the data as-is all the way through and letting the client decide whether to normalize before presentation.

Reply:
I don't think there are cases where containerd's `Normalize` will insert the _current_ architecture's info, right?  Doesn't `amd64` always means `linux/amd64`, even on Windows, for example?

Reply:
It does normalise the OS to be the "current" OS, so potentially show `darwin` or `macOS` on a Mac CLI; https://github.com/containerd/containerd/blob/36cc874494a56a253cd181a1a685b44b58a2e34a/platforms/database.go#L69-L80

```go
func normalizeOS(os string) string {
	if os == "" {
		return runtime.GOOS
	}
	os = strings.ToLower(os)

	switch os {
	case "macos":
		os = "darwin"
	}
	return os
}
```

Reply:
Ahhh, that's unfortunate (I guess that's a quirk of part of these APIs being designed for data storage / display and part of them being designed for querying, like the fallback vector code...) :disappointed:

Sounds like the best thing to do in this case then is just to display as-is -- if an image is missing the `os` value, we probably shouldn't inject/infer one for display, right?